apps/src/music/blockly/blocks/simple2.js

```diff
@@ -240,6 +249,120 @@ export const playSoundsSequential = {
       `
 };
 
+export const playSoundsRandom = {
+  definition: {
+    type: BlockTypes.PLAY_SOUNDS_RANDOM,
+    message0: 'play random',
+    args0: [],
+    message1: '%1',
+    args1: [
+      {
+        type: 'input_statement',
+        name: 'code'
+      }
+    ],
+    inputsInline: true,
+    previousStatement: null,
+    nextStatement: null,
+    style: 'flow_blocks',
+    tooltip: 'play sound randomly',
+    helpUrl: ''
+  },
+  generator: block => {
+    const resultArray = [];
+    let currentBlock = block.getInputTargetBlock('code');
+    while (currentBlock) {
+      const codeForBlock = getCodeForSingleBlock(currentBlock);
+      resultArray.push(codeForBlock);
+      currentBlock = currentBlock.getNextBlock();
+    }
+
+    const randomVar = Blockly.JavaScript.nameDB_.getDistinctName(
+      'random',
+      Blockly.Names.NameType.VARIABLE
+    );
+
+    let code = `var ${randomVar} = Math.floor(Math.random() * ${
+      resultArray.length
+    });`;
+
+    for (const [resultIndex, result] of resultArray.entries()) {
+      const lastMaybeSkipSoundVar = Blockly.JavaScript.nameDB_.getDistinctName(
+        '__lastMaybeSkipSound',
+        Blockly.Names.NameType.VARIABLE
+      );
+      const lastSkipSoundVar = Blockly.JavaScript.nameDB_.getDistinctName(
+        '__lastSkipSound',
+        Blockly.Names.NameType.VARIABLE
+      );
+      code += `
+        ${lastMaybeSkipSoundVar} = __maybeSkipSound;
+        ${lastSkipSoundVar} = __skipSound;
+        __maybeSkipSound = true;
+        __skipSound = __skipSound || ${randomVar} !== ${resultIndex};
+        ${result}
+        __maybeSkipSound = ${lastMaybeSkipSoundVar};
+        __skipSound = ${lastSkipSoundVar};
+        `;
+    }
+
+    return ` ProgramSequencer.playTogether();
+      ${code}
+      ProgramSequencer.endTogether();
+      `;
+  }
+};
+
+/**
+ * Generate code for the specified block but not following blocks.
+ * Adapted from this thread: https://groups.google.com/g/blockly/c/uXewhtr-mvM
+ * @param {Blockly.Block} block The block to generate code for.
+ * @return {string|!Array} For statement blocks, the generated code.
+ *     For value blocks, an array containing the generated code and an
+ *     operator order value.  Returns '' if block is null.
+ */
+function getCodeForSingleBlock(block) {
```

Comment:
Hmm interesting that this isn't a Blockly built-in. Might be worth splitting this out into a separate `blockUtils` file that we can reuse elsewhere?

Reply:
Good call; will do.